From d5bc99e0c5f34b3b49ee89a738425f125ec9cd25 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Thu, 20 Jun 2024 17:09:29 +0200
Subject: [PATCH 615/841] Revert "feat(stm32mp1): derivating MCE key from HUK"

This reverts commit fec05386e3931b85d34b3930b8b23d88f567b589.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 5b06f87a51cf673f33a922e59557a14ff8132971 Revert "feat(stm32mp1): derivating MCE key from HUK"]
Change-Id: I7e926f8c63106c42d652f9e9b7001ee5b99b7fce
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/388908
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 plat/st/stm32mp1/bl2_plat_setup.c           | 43 +++------------------
 plat/st/stm32mp1/include/stm32mp1_context.h |  4 +-
 plat/st/stm32mp1/platform.mk                |  4 +-
 plat/st/stm32mp1/stm32mp1_context.c         | 10 ++---
 plat/st/stm32mp1/stm32mp1_def.h             |  1 -
 5 files changed, 15 insertions(+), 47 deletions(-)

diff --git a/plat/st/stm32mp1/bl2_plat_setup.c b/plat/st/stm32mp1/bl2_plat_setup.c
index e14c85818..988397bc4 100644
--- a/plat/st/stm32mp1/bl2_plat_setup.c
+++ b/plat/st/stm32mp1/bl2_plat_setup.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015-2024, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2015-2023, Arm Limited and Contributors. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -24,9 +24,6 @@
 #include <drivers/st/stm32_mce.h>
 #endif
 #include <drivers/st/stm32_rng.h>
-#if STM32MP13
-#include <drivers/st/stm32_saes.h>
-#endif
 #include <drivers/st/stm32_uart.h>
 #include <drivers/st/stm32mp_reset.h>
 #include <drivers/st/stm32mp1_clk.h>
@@ -522,58 +519,30 @@ skip_console_init:
 }
 
 #if STM32MP13
-static int generate_mce_key_from_seed(uint8_t *mkey, uint8_t *seed, uint32_t seed_size){
-	int ret;
-	struct stm32_saes_context ctx;
-	uint8_t payload[MCE_KEY_SIZE_IN_BYTES] = {0U};
-
-	assert(seed_size <= MCE_KEY_SIZE_IN_BYTES);
-
-	ret = stm32_saes_driver_init();
-	if (ret != 0) {
-		return ret;
-	}
-
-	memcpy(payload, seed, seed_size); /* add seed in a block sized payload */
-
-	ret = stm32_saes_init(&ctx, false, STM32_SAES_MODE_ECB,
-			      STM32_SAES_KEY_DHU, NULL, MCE_KEY_SIZE_IN_BYTES, NULL, 0U);
-	if (ret != 0) {
-		return ret;
-	}
-
-	return stm32_saes_update(&ctx, true, payload, mkey, sizeof(payload));
-}
-
 static void prepare_encryption(void)
 {
 	uint8_t mkey[MCE_KEY_SIZE_IN_BYTES];
-	uint8_t seed[MCE_SEED_SIZE_IN_BYTES];
 
 	stm32_mce_init();
 
 #if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
-	if (stm32_rng_read(seed, MCE_SEED_SIZE_IN_BYTES) != 0) {
+	if (stm32_rng_read(mkey, MCE_KEY_SIZE_IN_BYTES) != 0) {
 		panic();
 	}
 #else /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 	if (stm32mp_is_wakeup_from_standby()) {
-		stm32mp1_pm_get_mce_seed_from_context(seed);
+		stm32mp1_pm_get_mce_mkey_from_context(mkey);
 		stm32_mce_reload_configuration();
 	} else {
-		/* Generate MCE master key seed from RNG */
-		if (stm32_rng_read(seed, MCE_SEED_SIZE_IN_BYTES) != 0) {
+		/* Generate MCE master key from RNG */
+		if (stm32_rng_read(mkey, MCE_KEY_SIZE_IN_BYTES) != 0) {
 			panic();
 		}
 
-		stm32mp1_pm_save_mce_seed_in_context(seed);
+		stm32mp1_pm_save_mce_mkey_in_context(mkey);
 	}
 #endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 
-	if (generate_mce_key_from_seed(mkey, seed, MCE_SEED_SIZE_IN_BYTES) != 0) {
-		panic();
-	}
-
 	if (stm32_mce_write_master_key(mkey) != 0) {
 		panic();
 	}
diff --git a/plat/st/stm32mp1/include/stm32mp1_context.h b/plat/st/stm32mp1/include/stm32mp1_context.h
index 42a00990a..9b815023c 100644
--- a/plat/st/stm32mp1/include/stm32mp1_context.h
+++ b/plat/st/stm32mp1/include/stm32mp1_context.h
@@ -19,8 +19,8 @@ bool stm32_pm_context_is_valid(void);
 void stm32_restore_ddr_training_area(void);
 uint32_t stm32_pm_get_optee_ep(void);
 #if STM32MP13
-void stm32mp1_pm_save_mce_seed_in_context(uint8_t *data);
-void stm32mp1_pm_get_mce_seed_from_context(uint8_t *data);
+void stm32mp1_pm_save_mce_mkey_in_context(uint8_t *data);
+void stm32mp1_pm_get_mce_mkey_from_context(uint8_t *data);
 void stm32mp1_pm_save_mce_region(uint32_t index, struct stm32_mce_region_s *config);
 void stm32mp1_pm_get_mce_region(uint32_t index, struct stm32_mce_region_s *config);
 #endif
diff --git a/plat/st/stm32mp1/platform.mk b/plat/st/stm32mp1/platform.mk
index d078710e4..ae0c711fe 100644
--- a/plat/st/stm32mp1/platform.mk
+++ b/plat/st/stm32mp1/platform.mk
@@ -239,8 +239,7 @@ PLAT_BL_COMMON_SOURCES  +=	drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c
 ifeq ($(STM32MP13),1)
 PLAT_BL_COMMON_SOURCES	+=	drivers/st/clk/clk-stm32-core.c				\
 				drivers/st/clk/clk-stm32mp13.c				\
-				drivers/st/crypto/stm32_rng.c				\
-				drivers/st/crypto/stm32_saes.c
+				drivers/st/crypto/stm32_rng.c
 else
 PLAT_BL_COMMON_SOURCES	+=	drivers/st/clk/stm32mp1_clk.c
 endif
@@ -258,6 +257,7 @@ endif
 ifeq (${TRUSTED_BOARD_BOOT},1)
 ifeq ($(STM32MP13),1)
 BL2_SOURCES		+=	drivers/st/crypto/stm32_pka.c
+BL2_SOURCES		+=	drivers/st/crypto/stm32_saes.c
 endif
 endif
 
diff --git a/plat/st/stm32mp1/stm32mp1_context.c b/plat/st/stm32mp1/stm32mp1_context.c
index 8684bdfe6..877c435cf 100644
--- a/plat/st/stm32mp1/stm32mp1_context.c
+++ b/plat/st/stm32mp1/stm32mp1_context.c
@@ -62,7 +62,7 @@ struct backup_data_s {
 	uint32_t bl2_code_end;
 	uint32_t bl2_end;
 #if STM32MP13
-	uint8_t mce_seed[MCE_SEED_SIZE_IN_BYTES];
+	uint8_t mce_mkey[MCE_KEY_SIZE_IN_BYTES];
 	struct stm32_mce_region_s mce_regions[MCE_IP_MAX_REGION_NB];
 #endif
 };
@@ -197,7 +197,7 @@ void stm32_restore_ddr_training_area(void)
 }
 
 #if STM32MP13
-void stm32mp1_pm_save_mce_seed_in_context(uint8_t *data)
+void stm32mp1_pm_save_mce_mkey_in_context(uint8_t *data)
 {
 	struct backup_data_s *backup_data;
 
@@ -209,12 +209,12 @@ void stm32mp1_pm_save_mce_seed_in_context(uint8_t *data)
 
 	clk_enable(BKPSRAM);
 
-	memcpy(backup_data->mce_seed, data, MCE_SEED_SIZE_IN_BYTES);
+	memcpy(backup_data->mce_mkey, data, MCE_KEY_SIZE_IN_BYTES);
 
 	clk_disable(BKPSRAM);
 }
 
-void stm32mp1_pm_get_mce_seed_from_context(uint8_t *data)
+void stm32mp1_pm_get_mce_mkey_from_context(uint8_t *data)
 {
 	struct backup_data_s *backup_data;
 
@@ -227,7 +227,7 @@ void stm32mp1_pm_get_mce_seed_from_context(uint8_t *data)
 
 	clk_enable(BKPSRAM);
 
-	memcpy(data, backup_data->mce_seed, MCE_SEED_SIZE_IN_BYTES);
+	memcpy(data, backup_data->mce_mkey, MCE_KEY_SIZE_IN_BYTES);
 
 	clk_disable(BKPSRAM);
 }
diff --git a/plat/st/stm32mp1/stm32mp1_def.h b/plat/st/stm32mp1/stm32mp1_def.h
index 9edce8f8f..45fc77e45 100644
--- a/plat/st/stm32mp1/stm32mp1_def.h
+++ b/plat/st/stm32mp1/stm32mp1_def.h
@@ -612,7 +612,6 @@ enum ddr_type {
 #if STM32MP13
 #define MCE_BASE			U(0x58001000)
 #define MCE_KEY_SIZE_IN_BYTES		U(16)
-#define MCE_SEED_SIZE_IN_BYTES		U(4)
 #endif
 
 /*******************************************************************************
-- 
2.47.2

